src/Aspire.Hosting.Azure.Functions/AzureFunctionsProjectResourceExtensions.cs

```diff
@@ -0,0 +1,150 @@
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting;
+
+namespace Aspire.Hosting.Azure;
+
+/// <summary>
+/// Extension methods for <see cref="AzureFunctionsProjectResource"/>.
+/// </summary>
+public static class AzureFunctionsProjectResourceExtensions
+{
+    /// <summary>
+    /// Add an Azure Functions project to the distributed application.
+    /// </summary>
+    /// <typeparam name="TProject"></typeparam>
+    /// <param name="builder"></param>
+    /// <param name="name"></param>
+    /// <returns></returns>
+    public static IResourceBuilder<AzureFunctionsProjectResource> AddAzureFunctionsProject<TProject>(this IDistributedApplicationBuilder builder, string name) where TProject : IProjectMetadata, new()
+    {
+        var projectDirectory = Path.GetDirectoryName(new TProject().ProjectPath)!;
+        var resource = new AzureFunctionsProjectResource(name, "func", projectDirectory);
+
+        // Add the default storage resource if it doesn't already exist.
+        var storage = builder.Resources.OfType<AzureStorageResource>().FirstOrDefault(r => r.Name == "azure-functions-default-storage");
+
+        if (storage is null)
+        {
+            storage = builder.AddAzureStorage("azure-functions-default-storage").RunAsEmulator().Resource;
+
+            builder.Eventing.Subscribe<BeforeStartEvent>((data, token) =>
+            {
+                var removeStorage = true;
+                // Look at all of the resources and if none of them use the default storage, then we can remove it.
+                // This is because we're unable to cleanly add a resource to the builder from within a callback.
+                foreach (var item in data.Model.Resources.OfType<AzureFunctionsProjectResource>())
+                {
+                    if (item.HostStorage == storage)
+                    {
+                        removeStorage = false;
+                    }
+
+                    // Before the resource starts, we want to apply the azure functions specific environment variables.
+                    // we look at all of the environment variables and apply the configuration for any resources that implement IResourceWithAzureFunctionsConfig.
+                    item.Annotations.Add(new EnvironmentCallbackAnnotation(static context =>
```

Comment:
Nice!

With this model, we could update more integrations packages to implement the `IResourceWithAzureFunctionsConfig` interface and they would light up in the Azure Functions resource without any additional updates on the `Aspire.Hosting.Azure.Functions` package.